mesh/ncmesh.cpp

```diff
@@ -4893,73 +4880,404 @@ void NCMesh::LoadVertexParents(std::istream &input)
+         face->attribute = attr;
+      }
+      else if (geom == Geometry::TRIANGLE)
+      {
+         input >> v1 >> v2 >> v3;
+         Face* face = faces.Get(v1, v2, v3);
+         face->attribute = attr;
+      }
+      else if (geom == Geometry::SEGMENT)
+      {
+         input >> v1 >> v2;
+         Face* face = faces.Get(v1, v1, v2, v2);
+         face->attribute = attr;
+      }
+      else
       {
-         MFEM_VERIFY(node.index() == node->p1, "invalid top-level vertex.");
-         MFEM_VERIFY(node->HasVertex(), "top-level vertex not found.");
-         MFEM_VERIFY(node->vert_index == node->p1, "bad top-level vertex index");
-         num_top_level = std::max(num_top_level, node->p1 + 1);
+         MFEM_ABORT("unsupported boundary element geometry: " << geom);
       }
    }
+}
+
+void NCMesh::PrintCoordinates(std::ostream &out) const
+{
+   int nv = coordinates.Size()/3;
+   out << nv << "\n";
+   if (!nv) { return; }
 
-   top_vertex_pos.SetSize(3*num_top_level);
-   for (int i = 0; i < num_top_level; i++)
+   out << spaceDim << "\n";
+   for (int i = 0; i < nv; i++)
    {
-      std::memcpy(&top_vertex_pos[3*i], mvertices[i](), 3*sizeof(double));
+      out << coordinates[3*i];
+      for (int j = 1; j < spaceDim; j++)
+      {
+         out << " " << coordinates[3*i + j];
+      }
+      out << "\n";
    }
 }
 
-int NCMesh::PrintElements(std::ostream &out, int elem, int &coarse_id) const
+void NCMesh::LoadCoordinates(std::istream &input)
 {
-   const Element &el = elements[elem];
-   if (el.ref_type)
+   int nv;
+   input >> nv;
+   if (!nv) { return; }
+
+   input >> spaceDim;
+
+   coordinates.SetSize(nv * 3);
+   coordinates = 0.0;
+
+   for (int i = 0; i < nv; i++)
    {
-      int child_id[8], nch = 0;
-      for (int i = 0; i < 8 && el.child[i] >= 0; i++)
+      for (int j = 0; j < spaceDim; j++)
       {
-         child_id[nch++] = PrintElements(out, el.child[i], coarse_id);
+         input >> coordinates[3*i + j];
+         MFEM_VERIFY(input.good(), "unexpected EOF");
       }
-      MFEM_ASSERT(nch == ref_type_num_children[(int) el.ref_type], "");
+   }
+}
+
+bool NCMesh::ZeroRootStates() const
+{
+   for (int i = 0; i < root_state.Size(); i++)
+   {
+      if (root_state[i]) { return false; }
+   }
+   return true;
+}
+
+void NCMesh::Print(std::ostream &out) const
+{
+   out << "MFEM nonconforming mesh v1.0\n\n"
+       "# NCMesh supported geometry types:\n"
+       "# TRIANGLE    = 2\n"
+       "# SQUARE      = 3\n"
+       "# TETRAHEDRON = 4\n"
+       "# CUBE        = 5\n"
+       "# PRISM       = 6\n";
+
+   out << "\ndimension\n" << Dim << "\n";
 
-      out << (int) el.ref_type;
-      for (int i = 0; i < nch; i++)
+   out << "\n# rank attr geom ref_type nodes/children";
+   out << "\nelements\n" << elements.Size() << "\n";
+
+   for (int i = 0; i < elements.Size(); i++)
+   {
+      const Element &el = elements[i];
+      out << el.rank << " " << el.attribute << " ";
```

Comment:
Sorry to bother again, but I just noticed that the new storing mechanism does not sync element attributes properly beween the NCMesh elements ant the Mesh elements. Changing element attributes via `Mesh::SetAttribute` does not affect the elements stored in the NCMesh, causing the mentioned issue. 

This should be fixable either through changing  `Mesh::SetAttribute` to also change the corresponding element attribute in the NCMesh element set. Since I could not figure out how to find the corresponding NCMesh element in constant time, I also thought alternatively about changing the signature of `NCMesh::Print` to 
```cpp
void NCMesh::Print(std::ostream &out, Mesh& mesh) const
```
and changing to this line to something like
```cpp
if(el.index == -1)
{
    out << el.rank << " " << el.attribute << " ";
}
else 
{
    out << el.rank << " " << mesh.GetAttribute(el.index) << " ";

}
```
looks also like it might work out as a hotfix.

Best regards, 
Dennis

Reply:
You're right and I think we should go for the first solution: `Mesh::SetAttribute` should call `NCMesh::SetAttribute` which would do `elements[leaf_elements[index]].attribute = attribute` or something similar.